### ISSUE TITLE
Creating new array inside expression string - Execute extension method

### ISSUE THREAD

**Issue by Mat3oo** [zzzprojects/EntityFramework-Classic#60]
With this expressions string passed into Execute extension method:
```cs
Where(p => new []{ 1, 2, 3, 4 }.Contains(p.Id))
```
library loses the `new` keyword and replace it with `(IEnumerable<int>)` cast, so causing exception.

Example [repo](https://github.com/Mat3oo/EntityFramework-Classic_Count_example/tree/newArray) (_newArray branch_) replicating the problem.

Exception details:
- Message:
```
The LINQ expression 'DbSet<Element>
    .Where(e => e.Name == "test")
    .Where(e => e.Active)
    .Where(e => (IEnumerable<int>)int[] { 1, 2, 3, 4, }
        .Contains(e.Id))' could not be translated. Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to either AsEnumerable(), AsAsyncEnumerable(), ToList(), or ToListAsync(). See https://go.microsoft.com/fwlink/?linkid=2101038 for more information.
```
- Stack trace:
```
at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.<VisitMethodCall>g__CheckTranslated|8_0(ShapedQueryExpression translated, <>c__DisplayClass8_0& )
   at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](Expression query, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass9_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQueryCore[TFunc](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1.GetEnumerator()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Z.Expressions.EvalLinq.Execute(String code, Boolean isQueryable, Expression exp, Object[] parameters)
   at System.Linq.QueryableDynamic.Execute[TValue](IQueryable source, String expression, Object parameter)
   at System.Linq.QueryableDynamic.Execute[TValue](IQueryable source, String expression)
   at EntityFramework_Classic_Count_example.Program.Main(String[] args) in *Removed by me*\EntityFramework-Classic_Count_example\EntityFramework-Classic_Count_example\Program.cs:line 21
```

**Comment by JonathanMagnan** [comment]
Hello @Mat3oo ,

Is there a reason why you don't create the list outside of the LINQ query and check in the LINQ if the list contains it?

Such as:

```csharp
var list = new List<int> { 1, 2, 3, 4};

DbSet<Element>
    .Where(e => e.Name == "test")
    .Where(e => e.Active)
    .Where(e => list.Contains(e.Id));
```

Best Regards,

Jon

**Comment by Mat3oo** [comment]
> 
> 
> Hello @Mat3oo ,
> 
> Is there a reason why you don't create the list outside of the LINQ query and check in the LINQ if the list contains it?
> 
> Such as:
> 
> ```cs
> var list = new List<int> { 1, 2, 3, 4};
> 
> DbSet<Element>
>     .Where(e => e.Name == "test")
>     .Where(e => e.Active)
>     .Where(e => list.Contains(e.Id));
> ```
> 
> Best Regards,
> 
> Jon

**Comment by Mat3oo** [comment]
I'm using urs Execute method:
```cs
DbSet<Element>
    .Execute<object>(
    "Where(e => e.Name == \"test\")" +
    "Where(e => e.Active)" +
    "Where(e => list.Contains(e.Id))");
```
and in this case it's not possible. In the repo from first comment is nice example.
*Missclick Close

**Comment by JonathanMagnan** [comment]
You can pass it in parameter such as:

```csharp
var list = new List<int> { 1, 2, 3, 4};

DbSet<Element>
    .Execute<object>(
    "Where(e => e.Name == \"test\")" +
    "Where(e => e.Active)" +
    "Where(e => list.Contains(e.Id))", new { list });
```

**Comment by Mat3oo** [comment]
> You can pass it in parameter

I know, but not the case for me. I passing string between services, so I just want to use `new [] {}` inplace.

**Comment by JonathanMagnan** [comment]
Hello @Mat3oo ,

Just to let you know that we continue to work on both issues.

We made some progression last week and find out why this happening (due to a convert on our side for using the `Contains` method).

Everything work without EF since this `CAST` doesn't cause any problem. However, in EF, this additional `CAST` cannot be translated in SQL which throw the error.

**Comment by JonathanMagnan** [comment]
Hello @Mat3oo ,

A new version has been released today.

We fixed this issue and the C# Eval Library will no longer try to `CAST` the array into `IEnumerable` which was causing the error in Entity Framework.

Let me know if everything works as expected.

Best Regards,

Jon

**Comment by Mat3oo** [comment]
It's seems with Z.Expressions.Eval v4.0.18 the problem `new []` is resolved, and also #59 working now as expected, gj 👍🏼.

**Comment by JonathanMagnan** [comment]
Great Mat3oo!

Don't hesitate to contact me for further assistance. 

Best regards, 

Jon
